data_structures/read_group.wdl

```diff
@@ -0,0 +1,386 @@
+            echo "~{if defined(read_group.SM) then ' SM:~{read_group.SM}' else ''}"
+        } > out.txt
+    >>>
+
+    output {
+        String stringified_read_group = read_string("out.txt")
+    }
+
+    runtime {
+        memory: "4 GB"
+        disk: "10 GB"
+        container: 'ghcr.io/stjudecloud/util:1.4.0'
+        maxRetries: 1
+    }
+}
+
+task get_ReadGroups {
+    meta {
+        description: "Gets read group information from a BAM file and writes it out as JSON which is converted to a WDL struct."
+        outputs: {
+            read_groups: "An array of strings containing read group information. If `format_for_star = true`, all found read groups are contained in one string (`read_groups[0]`). If `format_for_star = false`, each found @RG line will be its own entry in output array `read_groups`."
+        }
+    }
+
+    parameter_meta {
+        bam: "Input BAM format file to get read groups from"
+        modify_disk_size_gb: "Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB."
+    }
+
+    input {
+        File bam
+        Int modify_disk_size_gb = 0
+    }
+
+    Float bam_size = size(bam, "GiB")
+    Int disk_size_gb = ceil(bam_size) + 10 + modify_disk_size_gb
+
+    command <<<
+        set -euo pipefail
+        BAM="~{bam}" OUTFILE="read_groups.json" python - <<END
+import os  # lint-check: ignore
+import pysam  # lint-check: ignore
+import json  # lint-check: ignore
+sam = pysam.AlignmentFile(os.environ["BAM"], "rb")
+
+out_file = open(os.environ["OUTFILE"], "w")
+json.dump(sam.header.to_dict()["RG"], out_file)
+out_file.close()
+END
+    >>>
+
+    output {
+        Array[ReadGroup] read_groups = read_json("read_groups.json")
+    }
+
+    runtime {
+        memory: "4 GB"
+        disk: "~{disk_size_gb} GB"
+        container: 'quay.io/biocontainers/pysam:0.22.0--py38h15b938a_1'
+        maxRetries: 1
+    }
+}
+
+task validate_ReadGroup {
+    meta {
+        description: "Validate a ReadGroup struct's fields are defined"
+        outputs: {
+            check: "Dummy output to indicate success and enable call-caching"
+        }
+    }
+
+    parameter_meta {
+        read_group: "ReadGroup struct to validate"
+        required_fields: "Array of read group fields that must be defined"
+        restrictive: "If true, run a less permissive validation of field values. Otherwise, check against SAM spec-defined values."
+    }
+
+    input {
+        ReadGroup read_group
+        Array[String] required_fields = ["ID"]
+        Boolean restrictive = true
+    }
+
+    # The SAM spec allows any printable ASCII character in header fields.
+    String sam_spec_pattern = "[\\ -~]+"
+    # We have the opinion that is too permissive for ID and SM.
+    String id_pattern = if restrictive then "id" else sam_spec_pattern
+    String sample_pattern = if restrictive then "sample.?" else sam_spec_pattern
```

Comment:
I'm confused by what these patterns are meant to be. Doesn't this say the only acceptable `ID` value is the literal `id`? And Similarly that all `SM` values must start with the literal `sample`? That doesn't seem like what you intended.

Reply:
Below I test with 
```if [[ "~{read_group.SM}" =~ ^~{sample_pattern}$ ]]```
So for the sample test it tests for exactly the string `sample` plus one optional character. So `sample`, `sample1`, etc. would be excluded, but not `sample_1`. We could get overly complicated, but I was attempting to write a basic check that we could enhance over time if necessary, without tripping up too many users.

Reply:
That makes more sense but I'm still confused. Don't all these checks need to be negated?
```
        if [[ "~{read_group.ID}" =~ ^~{id_pattern}$ ]]
        then
            >&2 echo "ID must match pattern ~{id_pattern}"
            error=1
        fi
```
The first line will eval to true only when `read_group.ID` is the literal value `id` and then the string `ID must match pattern id` gets printed to STDERR. The ID val _does_ match the pattern.

Personally I think we should restrict to alphanumeric characters (plus some separators?) in ID or at least disbar whitespace. Same for SM?

Reply:
OK, I updated the checks. I always check for `sam_spec_pattern`. If `restrictive = true`, then `ID` disallows the literal string `id` and anything with a `space` and `SM` disallows the literal string `sample` with one or fewer additional characters (`sample.?`) and anything with a `space`.

I have three `String`s `id_pattern`, `sample_pattern`, and `restrictive_pattern` that contain the expressions. So if we decide to add additional patterns, we should be able to do that in the future fairly easily.